source/MRViewer/MRUIRectAllocator.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include "exports.h"
+#include "MRMesh/MRBox.h"
+#include "MRMesh/MRHash.h"
+#include "MRMesh/MRphmap.h"
+#include "MRMesh/MRVector2.h"
+
+#include <imgui.h>
+
+namespace MR::UI
+{
+
+// A generic rect allocator.
+// Given a set of rects and a target rect, finds the closest free rect next to the target (using dijkstra's algorithm).
+// The class can be reused multiple times to avoid repeated heap allocation, it doesn't store any state other than memory pools.
+// For optimal results you should bring your own AABB tree (see `FindPotentiallyOverlappingRects`) below,
+// but in simple cases you can get away with a flat list.
+class RectAllocator
+{
+public:
+    MRVIEWER_API RectAllocator();
```

Comment:
should it have virtual destructor?

Reply:
Shrug. I wouldn't add it until it's actually needed, since there are no other virtual functions.